kernel/names.mli

```diff
@@ -307,6 +307,60 @@ module KNset  : CSig.SetS with type elt = KerName.t
 module KNpred : Predicate.S with type elt = KerName.t
 module KNmap  : Map.ExtS with type key = KerName.t and module Set := KNset
 
+(** {6 Signature for quotiented names} *)
+
+module type EqType =
+sig
+  type t
+  val compare : t -> t -> int
+  val equal : t -> t -> bool
+  val hash : t -> int
+end
+
+module type QNameS =
+sig
+  type t
+  (** A type of reference that implements an implicit quotient by containing
+      two different names. The first one is the user name, i.e. what the user
+      sees when printing. The second one is the canonical name, which is the
+      actual absolute name of the reference.
+
+      This mechanism is fundamentally tied to the module system of Coq. Functor
+      application and module inclusion are the typical ways to introduce names
+      where the canonical and user components differ. In particular, the two
+      components should be undistinguishable from the point of view of typing,
+      i.e. from a "kernel" ground. This aliasing only makes sense inside an
+      environment, but at this point this notion is not even defined so, this
+      dual name trick is fragile. One has to ensure many invariants when
+      creating such names, but the kernel is quite lenient when it comes to
+      checking that these invariants hold. (Read: there are soundness bugs
+      lurking in the module system.)
+
+      One could enforce the invariants by splitting the names and storing that
+      information in the environment instead, but unfortunately, this wreaks
+      havoc in the upper layers. The latter are infamously not stable by
+      syntactic equality, in particular they might observe the difference
+      between canonical and user names if not packed together.
+
+      For this reason, it is discouraged to use the canonical-accessing API
+      in the upper layers, notably the [CanOrd] module below. Instead, one
```

Comment:
In your overlay for elpi you use `CanOrd`. I only need to build an OCaml Map/Set of GlobRefs. I think I don't I want convertible names to be distinguished by these set/map. Is `CanOrd` appropriate? (Maybe my old code was wrong in this respect)

Reply:
> I think I don't I want convertible names to be distinguished by these set/map.

This is precisely the main problem with canonical names, that is, nobody ever thinks about what semantic they actually want for the equality. They just pick the default `equal` / `compare` functions because it is what is obviously exposed. It's not clear that you always want to preserve the quotient though.

Codewise, eventually `CanOrd` should be removed, but this will take time to spot all the uses of canonical-accessing functions. So, for now it's fine to use in in plugins. This PR is just there to make the dependency explicit.

Reply:
Hum, I don't follow you. If the other order is the right one for Elpi, why can't use it now and avoid an overlay tomorrow?
Did you try it and it failed somewhere?

In elpi a `gref` (`GlobRef.t` in coq) is an opaque, unique, name for a global constant/inductive/constructor. And I expose maps/sets of them. I probably can't avoid Coq printing "the same" name in two different ways, but at least I can make these sets/map not sensitive to module aliases and the like.

Reply:
(I can change it on my side after this PR, I'm just trying to understand if I should)

Reply:
> at least I can make these sets/map not sensitive to module aliases and the like.

In the abstract, I'd recommend using the user name. I think that the ability to quotient references in the user-facing world is an anti-feature that was never thoroughly designed and merely evolved out of code compatibility. Terms which mention different names shouldn't be α-equivalent to start with.

I hope that we should be able to implement a decent module system in Coq (i.e. applicative functors) in the style of "F-ing modules", with actual semantics defined in an extension of CIC + one Type inaccessible in the hierarchy (i.e. Typeω). In this context, the canonical name could be some complex expression, so there is even less reason for it to be α-equivalent to the user name. Since when do we consider conversion as part of the syntactic equality?

Reply:
I agree with all you say. It's just I don't really know what this quotienting is for.

I was putting it at the same level of `Notation foo := bar` (even if it is implemented in the kernel). I want my grefs to be quotiented over Notation. If this thing is more like a delta step, then I'm fine not quotienting grefs up to delta.

Maybe the comment about not using CanOrd should say that, as of today, there is no better option (so that does not even look for it). The sentence "Eventially all uses will be removed" made me try to do it now :-/